Blocks: Create a Live Schedule block

ryelle opened the pull request:
Add a new block called "Live Schedule", which can be used on the Day of Event template (or any other page). Combined with #241, this will let organizers have total control over the content of the Day of Event template.

This also removes the need for a separate build script/process in `mu-plugins/theme-templates`, that will be removed in a following PR. It does remove the self-updating "Latest Posts" section, but organizers can use the Latest Posts block if they want, and this reduces the repeat API calls. If more organizers want this functionality we can see about hooking into the Latest Posts block.

Fixes #127 – You can customize the heading text and level in the block UI.
See #130 – Addresses a few issues from this issue

**Screenshots**

In wp-admin, it only displays a placeholder (for now). The headings are customized in the sidebar. 

![Screen Shot 2019-09-27 at 6 46 16 PM](https://user-images.githubusercontent.com/541093/65806363-2fb05380-e157-11e9-9cc1-b5e133f96ba2.png)

No sessions upcoming:

![Screen Shot 2019-09-27 at 6 51 33 PM](https://user-images.githubusercontent.com/541093/65806545-05ab6100-e158-11e9-8253-fdfcd22c0965.png)

List of sessions (faked "now" to be 10am):

![Screen Shot 2019-09-27 at 6 56 07 PM](https://user-images.githubusercontent.com/541093/65806628-7e122200-e158-11e9-99be-4e0805ffd110.png)

-----

Still to do (in following PRs)

- Fix issues with time
  - If we're after the start of the last event for a day, it says nothing's happening (we could check if there were other sessions on this day, and put up a "done for the day" message?)
  - If we're before the first event on the second day, it considers the last session of the prev day "on now".
- Remove `stripTagsAndEncodeText` and support diacritics.
- Audit the imports from the frontend script so we're not loading too much to the page
- Add a better, interactive preview in the editor
- (as mentioned) Remove the JS code from day of theme-templates.

------

**To test**

- Create a few sessions that are happening now (or add a property to `$data['live-schedule']` called `nowOverride` which should be a fake "now" in miliseconds).
- Add the Live Schedule block to a page.
- View it on the frontend, it should show you the current and upcoming sessions.

ryelle committed:
Add new Live Schedule block

This uses the code from the day of event template to create a Live Schedule block

ryelle committed:
Add controls for the Headings content

ryelle committed:
Add frontend rendering script

ryelle committed:
Use the heading & level saved in attribute

ryelle committed:
Update styles

ryelle committed:
Move live schedule config object to the WordCampBlocks object

ryelle committed:
Use the existing method for registering blocks

ryelle committed:
Only load the Live Schedule block in dev environments

ryelle commented on public_html/wp-content/mu-plugins/blocks/source/blocks/live-schedule/block.js:
```diff
@@ -0,0 +1,72 @@
+			isFetching: true,
+		};
+		this.update = this.update.bind( this );
+		this.interval = setInterval( this.update, 60 * 1000 ); // 60 seconds in ms.
```
We could tweak this a little, have it only fetch the data from the API every 10min or so but update the schedule display every minute. The schedule probably isn't changing much during the event, a 5-10min wait for fresh data would probably be fine - but we would still want the display to update quickly so when a session ends, the next is immediately visible. Thoughts?

iandunn replied:
If the API endpoint is cached on at the server/network layer -- see https://github.com/WordPress/wordcamp.org/issues/204#issuecomment-535231721 -- then I think leaving it this way is fine, and simplest from an application perspective.

If not, though, then I agree that it'd be ideal to fetch new data every 5-10 min, but have the client update the interface every minute, regardless of whether new data has come in.

`MUST` - If the endpoint isn't cached at the network layer, then changing this to only fetch every 5+ min should probably be a blocker, to avoid a situation similar to the Tagregator bug at WCUS several years ago.

ryelle replied:
> If the API endpoint is cached on at the server/network layer -- see #204 (comment) -- then I think leaving it this way is fine, and simplest from an application perspective.

I think it's a good idea to cache this at the server side, and eventually see if we can get the `If-Modified-Since` headers working (since this would still require a server hit, it won't protect against the tagregator issue alone). Let's move any more API cache discussion over to #207.

In any case, I've updated this PR to fetch the API every 5 minutes, while still re-rendering every minute.

melchoyce commented:
An idea for a placeholder state, using obviously fake session data:

![image](https://user-images.githubusercontent.com/2846578/66056786-a9ed2900-e505-11e9-8926-0bfd80eb6f01.png)

ryelle commented:
A future improvement could be to add a setting to restrict to a single track, so this block could be used on a livestream page.

ryelle committed:
Split re-rendering and API fetch into two intervals

This lets us request the API every 5 minutes to refresh the session data, while still updating the schedule rapidly.

ryelle committed:
Remove exceptions for `require-jsdoc` rule

ryelle committed:
Split editor & front-end css

Since front-end.js is its own entrypoint, the CSS imported there is pulled into `live-schedule.css`, while CSS imported into `index.js` is pulled in the global blocks.css which is loaded as the editor_style for all WC blocks.

ryelle committed:
Use fake data for the display in the editor

ryelle commented:
Rebased to remove the eslint `require-jsdoc` exceptions , and added the editor preview – so now it looks like this in the editor:

<img width="300" alt="Screen Shot 2019-10-11 at 3 42 34 PM" src="https://user-images.githubusercontent.com/541093/66680200-f50de700-ec3d-11e9-92be-773020cbdc9a.png">

The data we use to populate that is in [`data/mock.js`.](https://github.com/WordPress/wordcamp.org/blob/update/blocks-live-schedule/public_html/wp-content/mu-plugins/blocks/source/blocks/live-schedule/data/mock.js) It doesn't pull any API data, since the site probably doesn't have any active sessions when someone is editing the page.
